Socket: don't disconnect Socket for unknown/unsupported socket options.

tmds opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/59055.

@antonfirsov @dotnet/ncl ptal.

tmds committed:
Socket: don't disconnect Socket for unknown/unsupported socket options.

tmds committed:
Check SocketErrorCode

tmds committed:
test: use valid socket levels

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -3690,30 +3690,38 @@ internal void SetToDisconnected()
             // If we already know the socket is disconnected
             // we don't need to do anything else.
-            if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(this, $"errorCode:{errorCode}");
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(this, $"errorCode:{errorCode},noDisconnect:{noDisconnect}");
```
```suggestion
            if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(this, $"errorCode:{errorCode}, noDisconnect:{noDisconnect}");
```

antonfirsov reviewed (commented):
Thanks! I have only cosmetic concerns, otherwise looks good.

antonfirsov commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs:
```diff
@@ -27,7 +27,7 @@ public void ReuseUnicastPort_CreateSocketGetOption()
                 else
                 {
-                    Assert.Throws<SocketException>(() => socket.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseUnicastPort));
+                    SocketException se = Assert.Throws<SocketException>(() => socket.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseUnicastPort));
```
Why to assign 'SocketException se' if the variable is unused?

tmds committed:
Check CI results when removing OperationNotSupported

tmds committed:
TryGetPlatformSocketOption: differentiate between unknown levels and unknown options.

tmds committed:
PR feedback

tmds commented:
@antonfirsov can you take another look? I've addressed your feedback.

I've also made some additional changes. These have to do with us being able to detect use of an unknown option name (as `ENOPROTOOPT`), but not of an unknown option level. For the latter, the `errno` depends on the platform but is ambiguous (e.g. `EINVAL` can also mean the value is not valid).

We no longer disconnect for `ENOPROTOOPT`, but we keep disconnecting for the other errors (so also for unknown levels).

tmds committed:
Fix Android test failure.

tmds committed:
Revert changes to pal_networking

antonfirsov reviewed (approved):
The current changes look good. 

> We no longer disconnect for `ENOPROTOOPT`, but we keep disconnecting for the other errors (so also for unknown levels).

I see you reverted the attempt to fix this. Shall we merge this PR as-is, and  maybe open a new issue to track/discuss possible enhancements?
